File: ios/Classes/PdController/PdMessenger/PdMessengerImpl.swift
//
//  PdMessengerImpl.swift
//  flutter_libpd
//
//  Created by flav on 30/05/2022.
//

import Foundation
import libpd_ios

public class PdMessengerImpl: PdMessenger {
    
    public init() {}
    
    private func ensureSuccessStatus(_ status: Int32) throws {
        if status != 0 { throw MessageError.messageFailed }
    }
    
    public func sendFloat(_ message: FloatMessage) throws {
        let status = PdBase.send(message.value, toReceiver: message.receiver)
        try ensureSuccessStatus(status)
    }
    
    public func sendBang(_ message: BangMessage) throws {
        let status = PdBase.sendBang(toReceiver: message.receiver)
        try ensureSuccessStatus(status)
    }
    
    public func sendSymbol(_ message: SymbolMessage) throws {
        let status = PdBase.sendSymbol(message.value, toReceiver: message.receiver)
        try ensureSuccessStatus(status)
    }
    
    public func sendList(_ message: ListMessage) throws {
        let status = PdBase.sendList(message.value, toReceiver: message.receiver)
        try ensureSuccessStatus(status)
    }
    
    public func sendMessage(_ message: ArgumentMessage) throws {
        let status = PdBase.sendMessage(message.value, withArguments: message.arguments, toReceiver: message.receiver)
        try ensureSuccessStatus(status)
    }
}

File: android/src/main/kotlin/com/ovaom/flutter_libpd/pd_controller/pd_messenger/PdMessengerImpl.kt
package com.ovaom.flutter_libpd.pd_controller.pd_messenger

import org.puredata.core.PdBase

class PdMessengerImpl: PdMessenger {

    private fun ensureSuccessStatus(status: Int) {
        if (status != 0) throw MessageError.MessageFailed
    }

    override fun sendFloat(message: FloatMessage) {
        val status = PdBase.sendFloat(message.receiver, message.value)
        ensureSuccessStatus(status)
    }

    override fun sendBang(message: BangMessage) {
        val status = PdBase.sendBang(message.receiver)
        ensureSuccessStatus(status)
    }

    override fun sendSymbol(message: SymbolMessage) {
        val status = PdBase.sendSymbol(message.receiver, message.value)
        ensureSuccessStatus(status)
    }

    override fun sendList(message: ListMessage) {
        val status = PdBase.sendList(message.receiver, *message.value.toTypedArray())
        ensureSuccessStatus(status)
    }

    override fun sendMessage(message: ArgumentMessage) {
        val status = PdBase.sendMessage(message.receiver, message.value, *message.arguments.toTypedArray())
        ensureSuccessStatus(status)
    }
}